crates/ruff_python_formatter/src/range.rs

```diff
@@ -0,0 +1,793 @@
+
+    TextRange::new(start, range.end() - trailing_whitespace_len)
+}
+
+/// Finds the deepest node (ideally a statement) that fully encloses `range`.
+///
+/// ## Eligible nodes
+/// The search is restricted to nodes that mark the start of a logical line to ensure
+/// formatting a range results in the same formatting for that logical line as when formatting the entire document.
+/// This property can't be guaranteed when supporting sub-expression formatting because
+/// a) Adding parentheses around enclosing expressions can toggle an expression from non-splittable to splittable,
+/// b) formatting a sub-expression has fewer split points than formatting the entire expressions.
+///
+/// ### Possible docstrings
+/// Strings that are suspected to be docstrings are excluded from the search to format the enclosing suite instead
+/// so that the formatter's docstring detection in [`FormatSuite`] correctly detects it and formats it as a docstrings.
+///
+/// ### One line bodies
+/// Single line bodies `if True: pass` are excluded from the search to format the enclosing suite instead because the formatter
+/// must run [`FormatClauseBody`] to determine if the body should be collapsed or not.
+///
+/// ### Incorrectly indented code
+/// Code that uses indentations that don't match the configured [`IndentStyle`] and [`IndentWidth`] are excluded from the search,
+/// because formatting such nodes on their own can lead to indentation mismatch with its sibling nodes.
+///
+/// ## Suppression comments
+/// The search ends when `range` falls into a suppressed range because there's nothing to format. It also avoids that the
+/// formatter formats the statement because it doesn't see the suppression comment of the enclosing node.
+///
+/// The implementation doesn't handle `fmt: ignore` suppression comments because the statement's formatting logic
+/// correctly detects the suppression comment and returns the statement text as is.
+fn find_enclosing_node<'ast>(
+    range: TextRange,
+    root: AnyNodeRef<'ast>,
+    context: &PyFormatContext<'ast>,
+) -> EnclosingNode<'ast> {
+    let mut visitor = FindEnclosingNode::new(range, context);
+
+    if visitor.enter_node(root).is_traverse() {
+        root.visit_preorder(&mut visitor);
+    }
+    visitor.leave_node(root);
+
+    visitor.closest
+}
+
+struct FindEnclosingNode<'a, 'ast> {
+    range: TextRange,
+    context: &'a PyFormatContext<'ast>,
+
+    /// The, to this point, deepest node that fully encloses `range`.
+    closest: EnclosingNode<'ast>,
+
+    /// Tracks if the current statement is suppressed
+    suppressed: Suppressed,
+}
+
+impl<'a, 'ast> FindEnclosingNode<'a, 'ast> {
+    fn new(range: TextRange, context: &'a PyFormatContext<'ast>) -> Self {
+        Self {
+            range,
+            context,
+            suppressed: Suppressed::No,
+            closest: EnclosingNode::Suppressed,
+        }
+    }
+}
+
+impl<'ast> PreorderVisitor<'ast> for FindEnclosingNode<'_, 'ast> {
```

Comment:
Do you know why we use visitors instead of "regular" recursion? (Is it to avoid recursion causing stack overflows?)

Reply:
How do you mean? The visitors use recursion internally. I doubt that the visitor pattern (as used here) prevents stack overflows.

How would you write this with recursion without using a visitor? Would you use a queue with the nodes to visit and loop over it until it's empty (and push the node's children to the front while visiting?)